Makefile

```diff
@@ -1,4 +1,17 @@
 # ---------------------------------------------------------------------
+# Build configuration
+# ---------------------------------------------------------------------
+
+BUILD_DIRECTORY ?= release
+
+ifndef BUILD_DIRECTORY
```

Comment:
Nah, that's not quite what I meant (sorry for the confusion). I meant we should check that `$(dir $(BUILD_DIRECTORY))` is a directory that exists, so that if the user does e.g. `BUILD_DIRECTORY=/tmp/some/path make my-target` we should error if `/tmp/some` isn't a pre-existing directory.
Which seems like a reasonable thing to do IMHO?
(of course if `/tmp/some/path` already exists, that's just fine)

Hmmm, interesting thought: `BUILD_DIRECTORY=/tmp make clean` will try to delete all of `/tmp` which probably isn't what we want ;-)  The best fix for that might be to have `make clean` delete specifically-named files / dirs, rather than the whole `$(BUILD_DIRECTORY)`.